Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm

```diff
@@ -47,6 +51,128 @@
     completionHandler(enabledRulesets);
 }
 
+void WebExtensionContext::loadDeclarativeNetRequestRulesetStateFromStorage()
+{
+    NSDictionary *savedRulesetState = objectForKey<NSDictionary>(m_state, declarativeNetRequestRulesetStateKey);
+    for (NSString *savedIdentifier in savedRulesetState) {
+        for (auto& ruleset : extension().declarativeNetRequestRulesets()) {
+            if (ruleset.rulesetID == String(savedIdentifier)) {
+                ruleset.enabled = objectForKey<NSNumber>(savedRulesetState, savedIdentifier).boolValue;
+                break;
+            }
+        }
+    }
+}
+
+void WebExtensionContext::saveDeclarativeNetRequestRulesetStateToStorage(NSDictionary *rulesetState)
+{
+    NSDictionary *savedRulesetState = objectForKey<NSDictionary>(m_state, declarativeNetRequestRulesetStateKey);
+    NSMutableDictionary *updatedRulesetState = savedRulesetState ? [savedRulesetState mutableCopy] : [NSMutableDictionary dictionary];
+
+    [updatedRulesetState addEntriesFromDictionary:rulesetState];
+
+    [m_state setObject:[updatedRulesetState copy] forKey:declarativeNetRequestRulesetStateKey];
+    writeStateToStorage();
+}
+
+void WebExtensionContext::clearDeclarativeNetRequestRulesetState()
+{
+    [m_state removeObjectForKey:declarativeNetRequestRulesetStateKey];
+}
+
+WebExtensionContext::DeclarativeNetRequestValidatedRulesets WebExtensionContext::declarativeNetRequestValidateRulesetIdentifiers(const Vector<String>& rulesetIdentifiers)
+{
+    WebExtension::DeclarativeNetRequestRulesetVector validatedRulesets;
+
+    WebExtension::DeclarativeNetRequestRulesetVector rulesets = extension().declarativeNetRequestRulesets();
+
+    for (auto identifier : rulesetIdentifiers) {
+        bool found = false;
+        for (auto ruleset : rulesets) {
+            if (ruleset.rulesetID == identifier) {
+                validatedRulesets.append(ruleset);
+                found = true;
+                break;
+            }
+        }
+
+        if (!found)
+            return { std::nullopt, toErrorString(@"declarativeNetRequest.updateEnabledRulesets", nil, @"Failed to apply rules. Invalid ruleset id: %@.", (NSString *)identifier) };
```

Comment:
`@"declarativeNetRequest.updateEnabledRulesets()"`. Also the `NSString` cast might not be needed here.

Reply:
Added the parens, but this cast was needed:

/Users/brian_weinstein/OpenSource/Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm:100:156 Cannot pass object of non-trivial type 'WTF::String' through variadic function; call will abort at runtime